students/jack_anderson/lesson09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,227 @@
+        """
+        self._donations = sum(self.donations)
+        return self._donations
+
+    @property
+    def num_donations(self):
+        """
+        Property to count the number of donations made for a donor
+        :return: number of donations in donations list for donor
+        """
+        self._donations = len(self.donations)
+        return self._donations
+
+    @property
+    def avg_donation(self):
+        """
+        Property to return the avg donation made
+        :return: sum of donations / number of donations made
+        """
+        self._donations = self.sum_donations / self.num_donations
+        return self._donations
+
+
+    def report_template(self):
+        """
+        Returns text string of donor in a formatted layout containing name, total donations, num of donations made,
+        and the avg amount donated.
+        """
+        d = dict(name=self.name, total=self.sum_donations,count=self.num_donations, avg=self.avg_donation, width=10)
+        template = '{name:<21}\t$ {total:>{width}.2f}\t{count:^{width}}\t$ {avg:>{width}.2f}'.format(**d)
+        return template
+
+
+
+    def thanks_template(self):
+        """
+        Returns a 'thank you' style text string with donor name and total donations made for donor
+        """
+        d = dict(name=self.name, amount=self.sum_donations)
+        template =  "Hello {name},\nThank you for your gifts totaling ${amount:.2f}! \n" \
+                    "We will use your gift to help with costs for our upcoming play!\n" \
+                    "Thank you for giving!\nBest Regards,The Blanchford Community Center!".format(**d)
+        return template
+
+
+    def create_email(self, directory):
+        """Action to save a txt file of email to directory containing todays date
+        :param directory: name of directory to store emails
+        :return: text file of email to send to donor
+        """
+        self.date = date.today()
+        z = "{name}".format(name=self.name).replace(" ", "_")
+        now = "{date}".format(date=date.today()).replace("-", "_")
+        dir = directory
+        template = "{template}".format(template=self.thanks_template())
+        try:
+            with open('{dir}/{name}_{date}.txt'.format(dir=dir, name=z, date=now), 'w') as f:
+                f.write(template)
+        except TypeError:
+            raise TypeError
+
+
+
+class DonorCollection():
+
+    def __init__(self):
+        # Load donors info from donors.pckl file, if no file exists, create default file.
+        # Set self sorted dict variable
+        try:
+            with open('donors.pckl', 'rb') as f:
+                self.donors_dict = pickle.load(f)
+                f.close()
+        except (FileNotFoundError, AttributeError):
+            self.create_sample_list()
+            self.__init__()
+        self.sorted_dict = sorted(self.donors_dict.items(), key=lambda x: (sum(x[1]), x[0]), reverse=True)
```

Comment:
your DonorCollection class should hold a reference to multiple Donor object (which would be the correct usage of two classes)

right now you're sometimes creating Donor objects in few places